File: packages/python/tests/conformance_report.py
"""Emit this package's cross-validator conformance report."""

from __future__ import annotations

import json
import sys
from collections.abc import Iterator
from pathlib import Path
from typing import Any

from openakb_validate import __version__, extract_citations, validate

__all__ = ["main"]

_REPO_ROOT = Path(__file__).resolve().parents[3]
_CONFORMANCE = _REPO_ROOT / "conformance"


def _case_dirs(kind: str) -> list[Path]:
    return sorted(path for path in (_CONFORMANCE / kind).iterdir() if path.is_dir())


def _read_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def _codes(descriptor: object, *, strict: bool = False) -> list[str]:
    return sorted(validate(descriptor, strict=strict).codes)


def _cases() -> Iterator[tuple[str, dict[str, object]]]:
    for case_dir in _case_dirs("valid"):
        descriptor = _read_json(case_dir / "openakb.json")
        yield (
            f"valid/{case_dir.name}",
            {
                "lenient": _codes(descriptor),
                "strict": _codes(descriptor, strict=True),
            },
        )

    for case_dir in _case_dirs("invalid"):
        descriptor = _read_json(case_dir / "openakb.json")
        yield (
            f"invalid/{case_dir.name}",
            {
                "lenient": _codes(descriptor),
                "strict": _codes(descriptor, strict=True),
            },
        )

    for case_dir in _case_dirs("forward-compat"):
        descriptor = _read_json(case_dir / "openakb.json")
        yield (
            f"forward-compat/{case_dir.name}",
            {
                "lenient": _codes(descriptor),
                "strict": _codes(descriptor, strict=True),
            },
        )

    for case_dir in _case_dirs("content"):
        markdown = (case_dir / "content.md").read_text(encoding="utf-8")
        yield (
            f"content/{case_dir.name}",
            {"citations": [list(citation.ids) for citation in extract_citations(markdown)]},
        )


def main() -> int:
    report = {
        "implementation": "openakb-validate-python",
        "version": __version__,
        "spec_major": 1,
        "fixtures": dict(_cases()),
    }
    json.dump(report, sys.stdout, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/rust/examples/conformance_report.rs
//! Emits this crate's observed cross-validator conformance report.

use std::{
    collections::BTreeMap,
    fs,
    io::Write as _,
    path::{Path, PathBuf},
};

use anyhow::{Context, Result};
use openakb_validate::{Mode, extract_citations, validate};
use serde::Serialize;
use serde_json::Value;

#[derive(Debug, Serialize)]
struct Report {
    implementation: &'static str,
    version: &'static str,
    spec_major: u8,
    fixtures: BTreeMap<String, FixtureReport>,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum FixtureReport {
    Descriptor {
        lenient: Vec<String>,
        strict: Vec<String>,
    },
    Content {
        citations: Vec<Vec<String>>,
    },
}

fn main() -> Result<()> {
    let conformance_root = repo_root().join("conformance");
    let mut fixtures = BTreeMap::new();

    collect_descriptor_fixtures(&conformance_root, "valid", &mut fixtures)?;
    collect_descriptor_fixtures(&conformance_root, "invalid", &mut fixtures)?;
    collect_descriptor_fixtures(&conformance_root, "forward-compat", &mut fixtures)?;
    collect_content_fixtures(&conformance_root, &mut fixtures)?;

    let report = Report {
        implementation: "openakb-validate-rust",
        version: env!("CARGO_PKG_VERSION"),
        spec_major: 1,
        fixtures,
    };

    let stdout = std::io::stdout();
    let mut handle = stdout.lock();
    serde_json::to_writer_pretty(&mut handle, &report).context("write report JSON")?;
    handle.write_all(b"\n").context("write trailing newline")?;
    Ok(())
}

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("../..")
}

fn collect_descriptor_fixtures(
    conformance_root: &Path,
    kind: &str,
    fixtures: &mut BTreeMap<String, FixtureReport>,
) -> Result<()> {
    for case_dir in case_dirs(conformance_root, kind)? {
        let case_name = case_dir_name(&case_dir)?;
        let descriptor = read_json(&case_dir.join("openakb.json"))?;
        fixtures.insert(
            format!("{kind}/{case_name}"),
            FixtureReport::Descriptor {
                lenient: codes(&descriptor, Mode::Lenient),
                strict: codes(&descriptor, Mode::Strict),
            },
        );
    }
    Ok(())
}

fn collect_content_fixtures(
    conformance_root: &Path,
    fixtures: &mut BTreeMap<String, FixtureReport>,
) -> Result<()> {
    for case_dir in case_dirs(conformance_root, "content")? {
        let case_name = case_dir_name(&case_dir)?;
        let markdown = fs::read_to_string(case_dir.join("content.md"))
            .with_context(|| format!("read content fixture {case_name}"))?;
        let citations = extract_citations(&markdown)
            .into_iter()
            .map(|citation| citation.ids)
            .collect();
        fixtures.insert(
            format!("content/{case_name}"),
            FixtureReport::Content { citations },
        );
    }
    Ok(())
}

fn case_dirs(conformance_root: &Path, kind: &str) -> Result<Vec<PathBuf>> {
    let dir = conformance_root.join(kind);
    let mut paths = fs::read_dir(&dir)
        .with_context(|| format!("read fixture directory {}", dir.display()))?
        .map(|entry| {
            let entry = entry.with_context(|| format!("read entry in {}", dir.display()))?;
            Ok(entry.path())
        })
        .collect::<Result<Vec<_>>>()?;
    paths.retain(|path| path.is_dir());
    paths.sort();
    Ok(paths)
}

fn case_dir_name(case_dir: &Path) -> Result<String> {
    Ok(case_dir
        .file_name()
        .context("fixture directory has no final path component")?
        .to_string_lossy()
        .into_owned())
}

fn read_json(path: &Path) -> Result<Value> {
    let text = fs::read_to_string(path).with_context(|| format!("read {}", path.display()))?;
    serde_json::from_str(&text).with_context(|| format!("parse {}", path.display()))
}

fn codes(descriptor: &Value, mode: Mode) -> Vec<String> {
    validate(descriptor, mode)
        .codes()
        .into_iter()
        .map(|code| code.as_str().to_owned())
        .collect()
}
